demo_modules/wind/particle_field.js

```diff
@@ -0,0 +1,119 @@
+/* Copyright 2017 Google Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// Most of this code borrowed or derived from the awesome weather visualization
+// at https://earth.nullschool.net and its open source code:
+// https://github.com/cambecc/earth.
+
+const color = require('demo_modules/wind/color');
+
+const INTENSITY_SCALE_STEP = 5;
+const MAX_INTENSITY = 20;
+const MAX_PARTICLE_AGE = 100;
+const PARTICLE_LINE_WIDTH = 1;
+const PARTICLE_MULTIPLIER = 5;
+const FADE_FILL_STYLE = "rgba(0, 0, 0, 0.97)";
+
+class ParticleField {
+  constructor(bounds, grid, vectorField, context) {
+    this.bounds = bounds;
+    this.grid = grid;
+    this.vectorField = vectorField;
+    this.context = context;
+
+    this.initializeColorBuckets();
+    this.initializeParticles();
+  }
+
+  initializeColorBuckets() {
+    // maxIntensity is the velocity at which particle color intensity is maximum
+    this.colorStyles = color.windIntensityColorScale(
+        INTENSITY_SCALE_STEP, MAX_INTENSITY);
+
+    this.buckets = this.colorStyles.map(function() { return []; });
+  }
+
+  initializeParticles() {
+    const particleCount = Math.round(this.bounds.width * PARTICLE_MULTIPLIER);
+    this.particles = [];
+    for (var i = 0; i < particleCount; i++) {
+      this.particles.push(this.vectorField.randomize({age: _.random(0, MAX_PARTICLE_AGE)}));
+    }
+  }
+
+  evolve() {
+    this.buckets.forEach((bucket) => {
+      bucket.length = 0;
+    });
+    this.particles.forEach((particle) => {
+      if (particle.age > MAX_PARTICLE_AGE) {
+        this.vectorField.randomize(particle).age = 0;
+      }
+      const x = particle.x;
+      const y = particle.y;
+      const v = this.vectorField.vector(x, y);
+
+      const m = v[2];
+      if (m === null) {
+        // particle has escaped the grid, never to return...
+        particle.age = MAX_PARTICLE_AGE;
```

Comment:
Usually, we would reset the position here as well....

Reply:
If I reset the position, here, it would just be reset above. If I don't set age, it will get drawn. With it all refactored, does this comment still make sense?